types/wlr_surface_layers.c

```diff
@@ -0,0 +1,287 @@
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_MARGIN;
+}
+
+void wlr_layer_surface_move_state(struct wlr_layer_surface_state *next,
+		struct wlr_layer_surface_state *state) {
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_INTERACTIVITY) {
+		state->input_types = next->input_types;
+		state->exclusive_types = next->exclusive_types;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_ANCHOR) {
+		state->anchor = next->anchor;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_EXCLUSIVE_ZONE) {
+		state->exclusive_zone = next->exclusive_zone;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_MARGIN) {
+		state->margin_horizontal = next->margin_horizontal;
+		state->margin_vertical = next->margin_vertical;
+	}
+
+	state->invalid |= next->invalid;
+	next->invalid = 0;
+}
+
+void wlr_layer_surface_get_position(struct wlr_layer_surface *layer_surface,
+		double *x, double *y) {
+	int width = layer_surface->surface->current->width;
+	int height = layer_surface->surface->current->height;
+	struct wlr_layer_surface_state *state = layer_surface->current;
+
+	int output_width, output_height;
+	wlr_output_effective_resolution(layer_surface->output, &output_width,
+		&output_height);
+
+	if (state->anchor & LAYER_SURFACE_ANCHOR_LEFT) {
+		*x = state->margin_horizontal;
+	} else if (state->anchor & LAYER_SURFACE_ANCHOR_RIGHT) {
+		*x = output_width - width - state->margin_horizontal;
+	} else {
+		*x = (double)(output_width - width) / 2;
+	}
+	if (state->anchor & LAYER_SURFACE_ANCHOR_TOP) {
+		*y = state->margin_vertical;
+	} else if (state->anchor & LAYER_SURFACE_ANCHOR_BOTTOM) {
+		*y = output_height - height - state->margin_vertical;
+	} else {
+		*y = (double)(output_height - height) / 2;
+	}
+}
+
+void wlr_layer_surface_configure(struct wlr_layer_surface *layer_surface,
+		int32_t width, int32_t height) {
+	layer_surface_send_configure(layer_surface->resource, width, height);
+}
+
+static void layer_surface_destroy(struct wlr_layer_surface *surface) {
+	wl_signal_emit(&surface->events.destroy, surface);
+	wl_list_remove(&surface->link);
+	wl_resource_set_user_data(surface->resource, NULL);
+	free(surface->current);
+	free(surface->pending);
+	free(surface);
+}
+
+static void layer_surface_resource_destroy(struct wl_resource *resource) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	if (surface != NULL) {
+		layer_surface_destroy(surface);
+	}
+}
+
+static const struct layer_surface_interface layer_surface_impl = {
+	.set_interactivity = layer_surface_set_interactivity,
+	.set_anchor = layer_surface_set_anchor,
+	.set_exclusive_zone = layer_surface_set_exclusive_zone,
+	.set_margin = layer_surface_set_margin,
+};
+
+static void handle_layer_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct wlr_layer_surface *surface =
+		wl_container_of(listener, surface, surface_commit_listener);
+	wlr_layer_surface_move_state(surface->pending, surface->current);
```

Comment:
There's probably no reason to put state moving into its own function since there are only two kinds of state and moving the state only happens here. I did that in the surface implementation because there are three states that need to be moved around.